### DISCUSSION TITLE
What is missing for ETM algorithms to work

### DISCUSSION THREAD

**Discussion by juchom** [sshnet/SSH.NET#1202] [Q&A] [answered]
Hello,

I'm facing an issue where a remote server has been updated and it only supports these two HMAC algorithms :
hmac-sha2-256-etm@openssh.com
hmac-sha2-512-etm@openssh.com

There are already some issues open for this, and I was wondering what is missing to have this working ?

Is it a matter of time for the developper ?
Is it something missing in the framework ?

Or anything else ?

And thanks for this great library and the work done on it 👍

**Comment by Rob-Hague** [comment, marked as answer]
It is just that nobody has contributed the changes. I have looked into it briefly.

Because the EtM (encrypt-then-MAC) algorithms change the way the packets are created, there would need to be some refactoring around here:

https://github.com/sshnet/SSH.NET/blob/1c7166a002d7633fe1595b3df4634ba0ef6e3138/src/Renci.SshNet/Session.cs#L1052-L1066

depending on whether the MAC algorithm is defined to be EtM or not.

More details at https://crypto.stackexchange.com/questions/202/should-we-mac-then-encrypt-or-encrypt-then-mac

**Reply by juchom** [reply]
Ok, thanks for the info.
